plugins/trace_api_plugin/trace_api_plugin.cpp

```diff
@@ -0,0 +1,229 @@
+               e.what());
+         return fce.to_detail_string();
+      } catch (...) {
+         fc::unhandled_exception e(
+               FC_LOG_MESSAGE(warn, "unknown: ",),
+               std::current_exception());
+         return e.to_detail_string();
+      }
+   }
+
+   void log_exception( const exception_with_context& e, fc::log_level level ) {
+      if( _log.is_enabled( level ) ) {
+         auto detail_string = to_detail_string(std::get<0>(e));
+         auto context = fc::log_context( level, std::get<1>(e), std::get<2>(e), std::get<3>(e) );
+         _log.log(fc::log_message( context, detail_string ));
+      }
+   }
+}
+
+namespace eosio {
+
+/**
+ * A common source for information shared between the extraction process and the RPC process
+ */
+struct trace_api_common_impl {
+   static void set_program_options(appbase::options_description& cli, appbase::options_description& cfg) {
+      auto cfg_options = cfg.add_options();
+      cfg_options("trace-dir", bpo::value<bfs::path>()->default_value("traces"),
+                  "the location of the trace directory (absolute path or relative to application data dir)");
+      cfg_options("slice-stride", bpo::value<uint32_t>()->default_value(10'000),
+                  "the number of blocks each \"slice\" of trace data will contain on the filesystem");
+   }
+
+   void plugin_initialize(const appbase::variables_map& options) {
+      auto dir_option = options.at("trace-dir").as<bfs::path>();
+      if (dir_option.is_relative())
+         trace_dir = app().data_dir() / dir_option;
+      else
+         trace_dir = dir_option;
+
+      slice_stride = options.at("slice-stride").as<uint32_t>();
+   }
+
+   // common configuration paramters
+   boost::filesystem::path trace_dir;
+   uint32_t slice_stride = 0;
+};
+
+/**
+ * Interface with the RPC process
+ */
+struct trace_api_rpc_plugin_impl {
+   trace_api_rpc_plugin_impl( const std::shared_ptr<trace_api_common_impl>& common )
+   :common(common) {}
+
+   static void set_program_options(appbase::options_description& cli, appbase::options_description& cfg) {
+      auto cfg_options = cfg.add_options();
+      cfg_options("trace-rpc-abi", bpo::value<vector<string>>()->composing(),
+                  "ABIs used when decoding trace RPC responses, specified as \"Key=Value\" pairs in the form <account-name>=<abi-def>\n"
+                  "Where <abi-def> can be:\n"
+                  "   a valid JSON-encoded ABI as a string\n"
+                  "   an absolute path to a file containing a valid JSON-encoded ABI\n"
+                  "   a relative path from `data-dir` to a file containing a valid JSON-encoded ABI\n"
+                  );
+   }
+
+   void plugin_initialize(const appbase::variables_map& options) {
+      data_handler = std::make_shared<abi_data_handler>([](const exception_with_context& e){
+         log_exception(e, fc::log_level::debug);
+      });
+
+      if( options.count("trace-rpc-abi") ) {
```

Comment:
If not found, this should log a BIG warning. Or maybe we even have another option that explicitly indicates you are not providing abis and assert if not specified.